hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java

```diff
@@ -110,7 +110,46 @@ protected Boolean initialValue() {
   @Unstable
   public static <T extends Writable> AsyncGet<T, IOException>
       getAsyncRpcResponse() {
-    return (AsyncGet<T, IOException>) ASYNC_RPC_RESPONSE.get();
+    CompletableFuture<Writable> responseFuture = ASYNC_RPC_RESPONSE.get();
```

Comment:
org.apache.hadoop.util.functional.FutureIO should have what you need here already, with `raiseInnerCause` designed to expand that excution exception in a lot more detail.

* I'd propose a followup to switch to it
* and if there is something missing which is broadl useful: add it